source/extensions/filters/http/oauth/filter.cc

```diff
@@ -0,0 +1,475 @@
+
+      const auto state =
+          absl::StrReplaceAll(query_parameters.at(queryParamsState()), unescapedReplacements());
+      Http::Utility::Url state_url;
+      if (!state_url.initialize(state, false)) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      // Avoid infinite redirect storm
+      if (absl::StartsWith(state_url.pathAndQueryParams(), config_->callbackPath())) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      Http::ResponseHeaderMapPtr response_headers{
+          Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+              {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))},
+               {Http::Headers::get().Location, state}})};
+      decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+    }
+
+    // Continue on with the filter stack.
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // If a bearer token is supplied as a header or param, we ingest it here and kick off the
+  // user resolution immediately. Note this comes after HMAC validation, so technically this
+  // header is sanitized in a way, as the validation check forces the correct Bearer Cookie value.
+  access_token_ = extractAccessToken(headers, Token);
+  if (!access_token_.empty()) {
+    found_bearer_token_ = true;
+    request_headers_ = &headers;
+    oauth_client_->asyncGetIdentity(access_token_);
+
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // If no access token and this isn't the callback URI, redirect to acquire credentials.
+  //
+  // The following conditional could be replaced with a regex pattern-match,
+  // if we're concerned about matching strictly `/_oauth`.
+  if (!absl::StartsWith(path_str, config_->callbackPath())) {
+    Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+        {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
+
+    // Construct the correct scheme. We default to https since this is a requirement for OAuth to
+    // succeed. However, if a downstream client explicitly declares the "http" scheme for whatever
+    // reason, we also use "http" when constructing our redirect uri to the authorization server.
+    auto scheme = Http::Headers::get().SchemeValues.Https;
+
+    const auto* scheme_header = headers.Scheme();
+    if ((scheme_header != nullptr &&
+         scheme_header->value().getStringView() == Http::Headers::get().SchemeValues.Http)) {
+      scheme = Http::Headers::get().SchemeValues.Http;
+    }
+
+    const std::string base_path = absl::StrCat(scheme, "://", host_);
+    const std::string callback_path = base_path + config_->callbackPath();
+    const std::string state_path = absl::StrCat(base_path, headers.Path()->value().getStringView());
+
+    const std::string escaped_redirect_uri =
```

Comment:
i think we might be able to not escape here and just escape the individual components in the factory: the only unknown input is the host header which should already be safe? That said escaping might be cheap enough in the happy case that it doesn't matter